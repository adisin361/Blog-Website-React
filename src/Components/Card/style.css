.card {
  border: 1px solid grey;
  height: 45vh;
  width: 20vw;
  font-family: Arial, Helvetica, sans-serif;
  margin-bottom: 2vh;
}

.image img {
  width: 100%;
  height: 25vh
}


.description-section {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 40%;
  padding: 8px;
}

.date-time {
  color: grey;
  display: flex;
  justify-content: space-between;
}

.heading {
  padding-top: 7px;
  font-size: x-large;
}

.meta-data {
  padding: 7px 0px;
  color: grey;
  font-size: medium;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.clap {
  display: flex;
  align-items: center;
}


.reaction img {
  height: 2.5vh;
}

.reaction {
  display: flex;
  justify-content: space-between;
}

.footer hr {
  margin-bottom: 7px;
}

.clap p {
  font-size: medium;
  color: grey;
  padding-left: 5px;
}

.footer img {
  transition: all 0.2s ease-in-out
}

.footer img:hover {
  transition-duration: 0.2s;
  transform: scale(1.2);
}