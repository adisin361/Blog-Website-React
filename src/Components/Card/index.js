import * as React from 'react';
import './style.css';
import clapSrc from '../../Assets/Icons/clapping.svg';
import heartSrc from '../../Assets/Icons/heart-black.svg';
import heartLikedSrc from '../../Assets/Icons/heart-red.svg';
import clappedSrc from '../../Assets/Icons/clapped.svg';

import { UPDATE_BLOG_DATA } from '../../Constants/apiEndPoints';
import { getFormattedDateFromUtcDate } from '../../utils/common';
import makeRequest from '../../utils/makeRequest';

const Card = ({ blogData }) => {
  // const imageSrc = require(`../../Assets/Images/${props.image}`);
  const [isLiked, setIsLiked] = React.useState(blogData.liked);
  const [isClapped, setIsClapped] = React.useState(true);
  const [clapCount, setClapCount] = React.useState(blogData.claps);
  const [error, setError] = React.useState();

  const handleLike = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(blogData.id), {
        data: { liked: !isLiked },
      });
      setIsLiked(!isLiked);
    }
    catch (e) {
      setError(e.message);
    }
  };

  const handleClapCount = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(blogData.id), {
        data: { claps: clapCount + 1 }
      });
      setClapCount(clapCount + 1);
    }

    catch (e) {
      setError(e.message);
    }
  };

  const handleClap = () => {
    setIsClapped(!isClapped);
  }

  return (
    <>
      <div className='card'>
        <div className='image'>
          <img src={blogData.image} alt='abstract' />
        </div>
        <div className='description-section'>
          <div>
            <div className='date-time'>
              <p>{getFormattedDateFromUtcDate(blogData.date)}</p>
              <p>{blogData.reading_time}</p>
            </div>

            <p className='heading'>{blogData.title}</p>
            <p className='meta-data'>{blogData.description}</p>
          </div>

          <div className='footer'>
            <center><hr></hr></center>
            <div className='reaction'>
              <div className='clap'>
                <img src={isClapped ? clapSrc : clappedSrc} alt='clapIcon' onClick={() => {
                  handleClap(); handleClapCount()
                }} />
                <p data-testid='clap-counter'>{clapCount}</p>
              </div>
              <img src={isLiked ? heartLikedSrc : heartSrc} alt='heartIcon' onClick={handleLike} />
            </div>
          </div>
        </div>
      </div>
    </>

  );
};

export default Card;