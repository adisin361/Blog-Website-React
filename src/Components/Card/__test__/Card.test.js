import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import Card from "..";


const blog =
{
  "date": "2nd Januray, 2018",
  "readingTime": "2 mins",
  "title": "The future of abstract art and the culture ...",
  "description": "Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...",
  "claps": 10,
  "liked": false,
  "image": "abstract.png"
}

describe("Card", () => {
  // it("should match the snapshot", () => {
  //   const { asFragment } = render(<Card {...blog} />);
  //   expect(asFragment()).toMatchSnapshot();
  // });

  it("should change color of the heart icon to red when clicked", () => {
    render(<Card {...blog} />);
    expect(screen.queryByAltText('heartIcon').src).toEqual('http://localhost/heart-black.svg');
    fireEvent.click(screen.queryByAltText('heartIcon'));
    expect(screen.queryByAltText('heartIcon').src).toEqual('http://localhost/heart-red.svg');
  });

  it("should change color of the clap icon to orange when clicked", () => {
    render(<Card {...blog} />);
    expect(screen.queryByAltText('clapIcon').src).toContain('clapping.svg');
    fireEvent.click(screen.queryByAltText('clapIcon'));
    expect(screen.queryByAltText('clapIcon').src).toEqual('clapped.svg');
  });

  // it("should increment the clap counter when clicked once", () => {
  // render(<Card {...blog} />);
  // console.log(screen.getByTestId('clap-counter').innerText);
  // expect(screen.getByTestId('clap-counter')).(blog.claps);
  // fireEvent.click(screen.getByTestId('clap-counter'));
  // expect(screen.getByTestId('clap-counter').innerText).toEqual(blog.claps + 1);

  // });
  it("should increment the clap counter when clicked once", () => {
    render(<Card {...blog} />);
    expect(screen.getByTestId('clap-counter').innerText).toEqual(blog.claps.toString());
    fireEvent.click(screen.getByTestId('clap-counter'));
    expect(screen.getByTestId('clap-counter').innerText).toEqual((blog.claps + 1).toString());
  });

});

