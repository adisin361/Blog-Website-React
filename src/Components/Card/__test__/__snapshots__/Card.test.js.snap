// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="image"
    >
      <img
        alt="abstract"
        src="abstract.png"
      />
    </div>
    <div
      class="description-section"
    >
      <div>
        <div
          class="date-time"
        >
          <p>
            2nd Januray, 2018
          </p>
          <p>
            2 mins
          </p>
        </div>
        <p
          class="heading"
        >
          The future of abstract art and the culture ...
        </p>
        <p
          class="meta-data"
        >
          Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...
        </p>
      </div>
      <div
        class="footer"
      >
        <center>
          <hr />
        </center>
        <div
          class="reaction"
        >
          <div
            class="clap"
          >
            <img
              alt="clapIcon"
              src="clapping.svg"
            />
            <p>
              10
            </p>
          </div>
          <img
            alt="heartIcon"
            src="heart-black.svg"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
